### PR TITLE
chore(deps): Update `tree-sitter` dependencies to versions supporting Node 20

### PR DESCRIPTION
**Pull request by cr7pt0gr4ph7**
chore(deps): Update `tree-sitter` dependencies to versions supporting Node 20

This PR upates the different `tree-sitter` dependencies to versions that are compatible with Node 19/20.

## Background

The previously used version `0.20.x` of `tree-sitter` (the exact version number depends on the specific `tree-sitter-*` sub-package) does not work on Node 19 and higher, leading to errors like the following:

```
Request initialize failed with message: bad export type for `tree_sitter_(tsx|csharp|typescript|...)_external_scanner_create`: undefined (Internal Error)
```

This broke e.g. the [Official Cucumber VS Code extension](https://github.com/cucumber/vscode) when VS Code updated to a newer version of Node 19 (and later Node 20) that [was not supported at the time by the `tree-sitter` library](https://github.com/tree-sitter/tree-sitter-typescript/issues/244), but [support has since been implemented](https://github.com/tree-sitter/tree-sitter-typescript/pull/294) and subsequently shipped in `tree-sitter` `v0.21.0` and above.

Should fix the `bad export type for ...` error once the updated version is included [`cucumber/language-server`](https://github.com/cucumber/language-server) and then [`cucumber/vscode`](https://github.com/cucumber/vscode) that has been reported at least a couple times:

- cucumber/language-server#102
- cucumber/vscode#234
- cucumber/vscode#243
- cucumber/vscode#247
- cucumber/vscode#155

## Sidenote on versioning & compatibility of `tree-sitter`
The versioning and release scheme of the `tree-sitter` libraries is a bit "messy": The `tree-sitter-typescript`/`tree-sitter-rust`/... sub-libraries have a strong dependency on the major version number of `tree-sitter`, such that versions of the component libraries that require different major versions of the `tree-sitter` library cannot be used together. But they are:
- Not released at the same time.
- Do not have an central versioning number scheme.
- Introduce incompatible dependency upgrades from one patch version to the next (e.g. `tree-sitter-rust@0.23.1` requires `tree-sitter@0.21.1` while `tree-sitter-rust@0.23.2` requires `tree-sitter@0.22.1`).

(though that does not strictly violate SemVer requirements due to still being a `0.x.y` development version).

The approach taken here was to look for the newest version of `tree-sitter` for which all sub-libraries have at least one compatible release (which came out to be `tree-sitter@0.21.1`) and to choose the newest version of each sub-library that still supports `tree-sitter@0.21.1` (which eliminated a few of the most recent releases).

# Related PRs

This PR subsumes/replaces #225 as well as the following automated PRs (which have build failures due to trying to update the different dependencies separately):

- #180
- #183
- #184
- #185
- #188 (`tree-sitter` updated to `0.21.1` instead of `0.22.1` due to compatibility requirements, see below)
- #189 (`tree-sitter-python` updated to `0.23.4` instead of `0.23.6` due to compatibility requirements, see below)
- #201
- #234

## Tests

Tested with a custom build of `cucumber/vscode` to work on my current VS Code version:

```
Version: 1.95.3
Commit: f1a4fb101478ce6ec82fe9627c43efbf9e98c813
Date: 2024-11-13T14:50:04.152Z
Electron: 32.2.1
ElectronBuildId: 10427718
Chromium: 128.0.6613.186
Node.js: 20.18.0
V8: 12.8.374.38-electron.0
OS: Linux x64 6.11.3-300.fc41.x86_64
```

## A note on "missing import symbols" errors

If there are issues with missing import symbols (namely `missing symbol 'abort' from module 'env'`) after this update, this is due to a (too old) version of `emscripten` being used to compile `tree-sitter` to WASM.

The solution in that case is to install a newer version of `emscripten`.
- `tree-sitter` [currently uses](https://github.com/tree-sitter/tree-sitter/blob/dbe8bbf4808c4cad889595c2054fd9de0d734521/cli/loader/emscripten-version) `emscripten v3.1.64`
- My development environment used `emscripten v3.1.74`

### CODE REVIEW

**Comment by cr7pt0gr4ph7** [comment]
@kieran-ryan Test failures should be fixed now (`npm test` completed successfully when I ran it on my machine just now).

**Comment by kieran-ryan** [comment]
> @kieran-ryan Test failures should be fixed now (`npm test` completed successfully when I ran it on my machine just now).

Thanks a bunch for the contribution first and foremost @cr7pt0gr4ph7 - and for a well-documented pull request; will be super to have this issue resolved! Will look to fully test and review soon as I can - as time allows off on the back of holidays. Checking, are you testing on Windows by any chance; or any other other detail you can provide on your runtime? Will try to replicate difference in test outcome between running locally and the pipeline. There are some disparities in testing with the project which require resolution at some point to improve the contributing experience.

Suspect failing due to instability of query specifiers in tree-sitter language implementations (at least very early versions pinned in the package) which may have changed since with these updates - see `tree-sitter-rust` #226 for reference of associated changes. May not be the case, though was anticipating query changes with these upgrades.

[Tree-sitter's live playground](https://tree-sitter.github.io/tree-sitter/7-playground.html) can be used to check the query and test data associated with a language-implementation for compatibility - compared to the older versions our queries were compatible with; though you may want to [run a local tree-sitter playground against the specific language implementation versions you are running locally](https://tree-sitter.github.io/tree-sitter/cli/playground.html) for accuracy as they may differ.

**Comment by cr7pt0gr4ph7** [comment]
> Checking, are you testing on Windows by any chance; or any other other detail you can provide on your runtime?

I'm running on Linux Bluefin (Fedora-based) inside a Docker container:

```
$ uname --all
Linux a8f8f4bfa59d 6.11.3-300.fc41.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Oct 10 19:18:36 UTC 2024 x86_64 GNU/Linux

$ node --version
v22.12.0
```

> Will try to replicate difference in test outcome between running locally and the pipeline. There are some disparities in testing with the project which require resolution at some point to improve the contributing experience.

I'll have a look at the failed tests.

> Suspect failing due to instability of query specifiers in tree-sitter language implementations (at least very early versions pinned in the package) which may have changed since with these updates - see `tree-sitter-rust` #226 for reference of associated changes. May not be the case, though was anticipating query changes with these upgrades.

The test failures which I could observe locally were partially due to syntax tree changes in `tree-sitter-csharp`. I'll have a look!

> [Tree-sitter's live playground](https://tree-sitter.github.io/tree-sitter/7-playground.html) can be used to check the query and test data associated with a language-implementation for compatibility - compared to the older versions our queries were compatible with; though you may want to [run a local tree-sitter playground against the specific language implementation versions you are running locally](https://tree-sitter.github.io/tree-sitter/cli/playground.html) for accuracy as they may differ.

Thanks for the pointer! I actually looked for a Tree-sitter playground, but Google did not provide any helpful results. The manual page you linked does not even appear on the first page of results...

**Comment by cr7pt0gr4ph7** [comment]
@kieran-ryan Regarding further testing: I'd probably suggest updating `cucumber/vscode` first, and then wait for some time to if any additional problems are reported. It can't get any more broken than it is right now (where the language server doesn't even start in VS Code).

File: .github/workflows/test-javascript.yml [hunk 1]
```diff
@@ -20,7 +20,7 @@ jobs:
           - macos-latest
           - ubuntu-latest
           - windows-latest
-        node-version: ['18.x']
+        node-version: ['20.x']
```
Comment (kieran-ryan) [inline .github/workflows/test-javascript.yml:23]: Are we aware whether there are additional compatible node versions? We may want to pin with the same if necessary to prevent incompatible installations

Perhaps could strategise a series of Node versions (see [cucumber-js](https://github.com/cucumber/cucumber-js/blob/711d902693271eefee96696f7d234cfad37cb5cc/.github/workflows/build.yaml#L23)), trigger a workflow run and analyse compatibility.

File: .github/workflows/release-npm.yml [hunk 2]
```diff
@@ -14,10 +14,10 @@ jobs:
       - uses: mymindstorm/setup-emsdk@v14
         with:
           # See https://github.com/tree-sitter/tree-sitter/blob/master/cli/emscripten-version
```
Comment (kieran-ryan) [inline .github/workflows/release-npm.yml:16]: [Previously pointed to the version](https://github.com/tree-sitter/tree-sitter/blob/v0.20.2/cli/emscripten-version), though [link is now dead](https://github.com/tree-sitter/tree-sitter/blob/master/cli/emscripten-version) so we may want to [update and use a permalink](https://github.com/tree-sitter/tree-sitter/blob/cd3d967f3cc1ab005e6f122f2cd034d0b02321ff/cli/loader/emscripten-version) - similarly applying to [reference in testworkflow.yml](https://github.com/cucumber/language-service/pull/235/files#diff-b1d8ed23b728461de1cd6c65454bb692f69d5e4de0a3e32f38e0ddec13103853R32)

**Review by kieran-ryan — CHANGES_REQUESTED** [review]
@cr7pt0gr4ph7, would you be aware of the [minimum supported version of VSCode](https://github.com/cucumber/vscode/blob/cc6d46729db39f6267faafbb724c484ffa593981/package.json#L8) following these changes; along with the range of supported node versions?

**Comment by kieran-ryan** [comment]
@cr7pt0gr4ph7, can for sure understand that perspective; would just be conscious of current users using the language server directly or through the VSCode extension that have pinned an earlier version of VSCode - as would autoupdate and could require promptly jumping on any fixes.

**Comment by cr7pt0gr4ph7** [comment]
> Would you be aware of the [minimum supported version of VSCode](https://github.com/cucumber/vscode/blob/cc6d46729db39f6267faafbb724c484ffa593981/package.json#L8) following these changes; along with the range of supported node versions?

I haven't modified anything related to communication with VS Code or upgraded any dependency beside `tree-sitter` and it's sub-libraries.

It therefore all comes down to the compatibility between `tree-siter` and emscripten version we use to compile to ASM on one side, and the Node.js versions included in different VS Code versions on the other.

- There [isn't an offical table](https://github.com/microsoft/vscode/issues/85939) of the Node.js versions shipped for each VS Code release, but utilizing the script from [`Sneezry/vscode-version-watcher`](https://github.com/Sneezry/vscode-version-watcher) (with some tweaks to make it run again) produces the table below.
- Both `emscripten v3.1.64` and `tree-sitter` unfortunately do not seem to have an official matrix of supported Node versions; running our tests in the different versions of Node should be enough to verify compatibility, though.

> Last update: 2025-01-02 21:02:51 GMT
>
> | VS Code | Electron | Node | Chrome |
> |:-------:|:--------:|:----:|:------:|
> | Latest | 32.2.7 | 20.18.1 | 128.0.6613.186 |
> | 1.96.2 | 32.2.6 | 20.18.1 | 128.0.6613.186 |
> | 1.96.1 | 32.2.6 | 20.18.1 | 128.0.6613.186 |
> | 1.96.0 | 32.2.6 | 20.18.1 | 128.0.6613.186 |
> | 1.95.3 | 32.2.1 | 20.18.0 | 128.0.6613.186 |
> | 1.95.2 | 32.2.1 | 20.18.0 | 128.0.6613.186 |
> | 1.95.1 | 32.2.1 | 20.18.0 | 128.0.6613.186 |
> | 1.95.0 | 32.2.1 | 20.18.0 | 128.0.6613.186 |
> | 1.94.2 | 30.5.1 | 20.16.0 | 124.0.6367.243 |
> | 1.94.1 | 30.5.1 | 20.16.0 | 124.0.6367.243 |
> | 1.94.0 | 30.5.1 | 20.16.0 | 124.0.6367.243 |
> | 1.93.1 | 30.4.0 | 20.15.1 | 124.0.6367.243 |
> | 1.93.0 | 30.4.0 | 20.15.1 | 124.0.6367.243 |
> | 1.92.2 | 30.1.2 | 20.14.0 | 124.0.6367.243 |
> | 1.92.1 | 30.1.2 | 20.14.0 | 124.0.6367.243 |
> | 1.92.0 | 30.1.2 | 20.14.0 | 124.0.6367.243 |
> | 1.91.1 | 29.4.0 | 20.9.0 | 122.0.6261.156 |
> | 1.91.0 | 29.4.0 | 20.9.0 | 122.0.6261.156 |
> | 1.90.2 | 29.4.0 | 20.9.0 | 122.0.6261.156 |
> | 1.90.1 | 29.4.0 | 20.9.0 | 122.0.6261.156 |
> | 1.90.0 | 29.4.0 | 20.9.0 | 122.0.6261.156 |
> | 1.89.1 | 28.2.8 | 18.18.2 | 120.0.6099.291 |
> | 1.89.0 | 28.2.8 | 18.18.2 | 120.0.6099.291 |
> | 1.88.1 | 28.2.8 | 18.18.2 | 120.0.6099.291 |
> | 1.88.0 | 28.2.8 | 18.18.2 | 120.0.6099.291 |
> | ... | ... | ... | ... |
>
> (Table was truncated for readability to only show the history up until the last known-good version 1.89 of VS Code)

> Would just be conscious of current users using the language server directly or through the VSCode extension that have pinned an earlier version of VSCode - as would autoupdate and could require promptly jumping on any fixes.

You're absolutely right; I hadn't considered that scenario...

File: .github/workflows/release-npm.yml (hunk 2 above)
Comment (cr7pt0gr4ph7) [reply .github/workflows/release-npm.yml:16]: I'd suggest the following:
```suggestion
          # See https://github.com/tree-sitter/tree-sitter/blob/v0.24.6/cli/loader/emscripten-version
          # (choose the Git tag that corresponds to the version of web-tree-sitter specified in package.json)
          # Should be kept in sync with .github/workflows/test-javascript.yml
```

File: .github/workflows/test-javascript.yml [hunk 3]
```diff
@@ -30,7 +30,7 @@ jobs:
         if: ${{ matrix.os != 'windows-latest' }}
         with:
           # See https://github.com/tree-sitter/tree-sitter/blob/master/cli/emscripten-version
```
Comment (cr7pt0gr4ph7) [inline .github/workflows/test-javascript.yml:32]: @kieran-ryan Same here:
```suggestion
          # See https://github.com/tree-sitter/tree-sitter/blob/v0.24.6/cli/loader/emscripten-version
          # (choose the Git tag that corresponds to the version of web-tree-sitter specified in package.json) 
          # Should be kept in sync with .github/workflows/release-npm.yml
```

**Comment by cr7pt0gr4ph7** [comment]
I just had a look at the new CI test results, which seem good so far. A few notes:

- The Windows tests always succeed because they don't even run the WASM tests... (because they don't install the `tree-sitter` CLI tool on Windows) 

  ```
  > @cucumber/language-service@1.6.0 prepare
  > husky && node scripts/build.js && cp node_modules/web-tree-sitter/tree-sitter.wasm dist

  Skipping compilation of tree-sitter wasms - node_modules\.bin\tree-sitter is not installed
  ```

- The 19.x tests fail because `@cucumber/cucumber` and the `@typescript-eslint/*` packages are marked as incompatible with Node 19 (probably because it is a non-LTS release that is already past its six month support window). Though VS Code skipped straight from 19.x to 20.x, so we would probably be fine with supporting only 18.x and 19.x.

So, in conclusion: It _should_ likely be running just fine in VS Code 1.82.0, which uses Node 18.15.0. I'm unfortunately not able to verify this locally with an old version of VS Code due to the way my system is set up.

**Comment by cr7pt0gr4ph7** [comment]
I updated `test-javascript.yaml` to run the tests on every current LTS Node release (`18.x`, `20.x`, `22.x`, `24.x`) because while VS Code uses Node 20.x right now, other clients might use newer versions.

Based on the successful tests on Node 18.x and 20.x (and provided that the tests for Node `22.x` and `24.x` also succeed), this should be ready for merge from my viewpoint.

File: .github/workflows/test-javascript.yml (hunk 1 above)
Comment (cr7pt0gr4ph7) [reply .github/workflows/test-javascript.yml]: I've updated `test-javascript.yml` to run tests on all current LTS releases of Node (`18.x`, `20.x`, `22.x`, `24.x`) :+1:

**Comment by cr7pt0gr4ph7** [comment]
As I discovered just now, Node 24.x does not even exist yet 😅 It is due for release around April/May of 2025.

Node 22.x seems to work according to the test, meaning that all current LTS releases (18.x, 20.x, 22.x) are supported.

**Comment by kieran-ryan** [comment]
> As I discovered just now, Node 24.x does not even exist yet 😅 It is due for release around April/May of 2025.
> 
> Node 22.x seems to work according to the tests, meaning that all current LTS releases (18.x, 20.x, 22.x) are now supported.

Awesome 🙌 Re-running workflow. Fantastic has full support - and running in CI. This was lacking previously for v20, which subsequently became an issue accordingly.

We might want to specify the `"engines"` and `"enginesTested"` within `package.json` (see [cucumber-js](https://github.com/cucumber/cucumber-js/blob/f0ef00126490a0824bfafa3d31bddbfae5082998/package.json#L212-L217)) for package managers to handle compatibility and user feedback. Suggest we also add `23.x` to the workflow as well (see [cucumber-js](https://github.com/cucumber/cucumber-js/blob/f0ef00126490a0824bfafa3d31bddbfae5082998/.github/workflows/build.yaml#L23)) to assess compatibility of the same.

Thanks @cr7pt0gr4ph7!!

**Comment by cr7pt0gr4ph7** [comment]
> We might want to specify the `"engines"` and `"enginesTested"` within `package.json` (see [cucumber-js](https://github.com/cucumber/cucumber-js/blob/f0ef00126490a0824bfafa3d31bddbfae5082998/package.json#L212-L217)) for package managers to handle compatibility and user feedback.

I didn't add it yet because I wasn't sure how much of a "guarantee" the `"engines"` and `"enginesTested"` properties actually communicate (as a sidenote, `"enginesTested"` does not seem to be an [official `package.json` field](https://docs.npmjs.com/cli/v11/configuring-npm/package-json), but I see its usefullness).

> Suggest we also add `23.x` to the workflow as well (see [cucumber-js](https://github.com/cucumber/cucumber-js/blob/f0ef00126490a0824bfafa3d31bddbfae5082998/.github/workflows/build.yaml#L23)) to assess compatibility of the same.

That's mainly a matter of project policy - `23.x` will become unsupported in ~4 months anyway. I've added it now.

**Comment by kieran-ryan** [comment]
@cr7pt0gr4ph7, would you be able to include a `CHANGELOG.md` entry that support has been `Added` for the specified Node versions; for inclusion within release notes? Integration testing I have performed has been successful; planning to test with different VSCode versions prior to merge.

**Comment by cr7pt0gr4ph7** [comment]
> @cr7pt0gr4ph7, would you be able to include a `CHANGELOG.md` entry that support has been `Added` for the specified Node versions; for inclusion within release notes?

Sure. Done :+1: Also slightly updated the PR title to better match the supported version range.

**Comment by neskk** [comment]
I've tested this locally (built it within WSL) and it works fine :)

**Comment by cr7pt0gr4ph7** [comment]
> I've tested this locally (built it within WSL) and it works fine 😄

Great! Looking forward to getting this merged :)

This is running in a language server outside of the main VS Code process, so as long as the LSP interface wasn't changed and it is compatible with the runtime environment (as shown by the CI), I don't expect any additional problems.

**Comment by johnDeSilencio** [comment]
Just wanted to chime in here and say that I have also been looking forward to this change getting merged!

I have an unconventional use case. I am trying to use `cucumber/language-server` inside of Neovim on NixOS. I can confirm that changes that @cr7pt0gr4ph7 has introduced do indeed fix the ```bad export type for `tree_sitter_...` ``` error that I was seeing before.

For any other NixOS + Neovim users that are interested in seeing how I did it, they are welcome to look at [the fork I made of `cucumber/language-server`](https://github.com/johnDeSilencio/language-server) that include @cr7pt0gr4ph7's changes and to check out [my NixOS config](https://github.com/johnDeSilencio/NicksOS/blob/main/modules/nixos/language-servers/default.nix#L17).

(Note that I had to commit a crime on the fork of `cucumber/language-server` and commit the `dist/` directory because I'm new to defining nix packages and that was the only way I could get everything to work. It works on my machine at least :smile: ) 

2025-03-01-151501_hyprshot
23
2025-03-01-151609_hyprshot

**Comment by ant0x64** [comment]
https://github.com/orgs/cucumber/discussions/2253
